File: @.PTM.℗™/pmo-gh/1.PIECE.MOD.OS_c&ph_04.01/!.PMO_c_00.01/pieces/apps/fuzzpet_app/fuzzpet/plugins/fuzzpet.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <unistd.h>
#include <time.h>
#include <pthread.h>
#include <sys/stat.h>
#include <errno.h>

#define MAX_LINE 256

// Standardized Key Codes
#define K_LEFT  1000
#define K_RIGHT 1001
#define K_UP    1002
#define K_DOWN  1003

// Map dimensions
#define MAP_ROWS 10
#define MAP_COLS 20

char map[MAP_ROWS][MAP_COLS + 1];
char last_key_display[64] = "None";
char current_response_key[64] = "default";

// --- TPM Helpers ---

int get_piece_state_int(const char* piece, const char* key) {
    char cmd[512];
    snprintf(cmd, sizeof(cmd), "./pieces/master_ledger/plugins/+x/piece_manager.+x %s get-state %s", piece, key);
    FILE *fp = popen(cmd, "r");
    if (!fp) return 0;
    char val[64];
    int result = 0;
    if (fgets(val, sizeof(val), fp)) {
        if (strcmp(val, "STATE_NOT_FOUND") != 0) result = atoi(val);
    }
    pclose(fp);
    return result;
}

void set_piece_state_int(const char* piece, const char* key, int val) {
    char cmd[512];
    snprintf(cmd, sizeof(cmd), "./pieces/master_ledger/plugins/+x/piece_manager.+x %s set-state %s %d > /dev/null 2>&1", piece, key, val);
    system(cmd);
}

void log_master(const char* type, const char* event, const char* piece, const char* source) {
    time_t rawtime;
    struct tm *timeinfo;
    char timestamp[100];
    time(&rawtime);
    timeinfo = localtime(&rawtime);
    strftime(timestamp, sizeof(timestamp), "%Y-%m-%d %H:%M:%S", timeinfo);
    
    FILE *ledger = fopen("pieces/master_ledger/master_ledger.txt", "a");
    if (ledger) {
        fprintf(ledger, "[%s] %s: %s on %s | Source: %s\n", timestamp, type, event, piece, source);
        fclose(ledger);
    }
}

// --- FuzzPet Logic ---

void sync_mirror() {
    int hunger = get_piece_state_int("fuzzpet", "hunger");
    int happiness = get_piece_state_int("fuzzpet", "happiness");
    int energy = get_piece_state_int("fuzzpet", "energy");
    int level = get_piece_state_int("fuzzpet", "level");
    int pos_x = get_piece_state_int("fuzzpet", "pos_x");
    int pos_y = get_piece_state_int("fuzzpet", "pos_y");

    FILE *f = fopen("pieces/apps/fuzzpet_app/fuzzpet/state.txt", "w");
    if (f) {
        fprintf(f, "hunger=%d\nhappiness=%d\nenergy=%d\nlevel=%d\npos_x=%d\npos_y=%d\nstatus=active\n",
                hunger, happiness, energy, level, pos_x, pos_y);
        fclose(f);
    }
}

void write_sovereign_view() {
    // 1. Sync the mirror first so others see the data (Parser relies on this!)
    sync_mirror();

    // 2. Get Last Response (Audit Reflection)
    char response[256] = "";
    char cmd[512];
    snprintf(cmd, sizeof(cmd), "./pieces/master_ledger/plugins/+x/piece_manager.+x fuzzpet get-response %s", current_response_key);
    FILE *rf = popen(cmd, "r");
    if (rf) {
        fgets(response, sizeof(response), rf);
        response[strcspn(response, "\n")] = 0;
        pclose(rf);
    }

    // 3. Compose Sovereign View (The "Stage")
    FILE *fp = fopen("pieces/apps/fuzzpet_app/fuzzpet/view.txt", "w");
    if (!fp) return;

    for (int y = 0; y < MAP_ROWS; y++) {
        fprintf(fp, "║  %-20s                                     ║\n", map[y]);
    }
    fprintf(fp, "╠═══════════════════════════════════════════════════════════╣\n");
    fprintf(fp, "║  [FUZZPET]: %-45s ║\n", response);
    fprintf(fp, "║  DEBUG [Last Key]: %-38s ║\n", last_key_display);
    
    fclose(fp);
    
    // SIGNALING NODE: Append to local marker to wake up the Parser (Option Z)
    FILE *marker = fopen("pieces/apps/fuzzpet_app/fuzzpet/view_changed.txt", "a");
    if (marker) {
        fputc('X', marker);
        fputc('\n', marker);
        fclose(marker);
    }
}

void save_map() {
    FILE *f = fopen("pieces/apps/fuzzpet_app/world/map.txt", "w");
    if (f) {
        for (int y = 0; y < MAP_ROWS; y++) {
            fprintf(f, "%.*s\n", MAP_COLS, map[y]);
        }
        fclose(f);
    }
}

void load_map() {
    FILE *f = fopen("pieces/apps/fuzzpet_app/world/map.txt", "r");
    if (f) {
        for (int y = 0; y < MAP_ROWS; y++) {
            if (fgets(map[y], MAP_COLS + 2, f)) { 
                map[y][strcspn(map[y], "\n\r")] = 0; 
            } else {
                map[y][0] = '\0'; 
            }
        }
        fclose(f);
    } else {
        for (int y = 0; y < MAP_ROWS; y++) {
            for (int x = 0; x < MAP_COLS; x++) map[y][x] = '#'; 
            map[y][MAP_COLS] = 0; 
        }
        save_map();
    }

    int initial_pos_x = get_piece_state_int("fuzzpet", "pos_x");
    int initial_pos_y = get_piece_state_int("fuzzpet", "pos_y");

    for (int y = 0; y < MAP_ROWS; y++) {
        for (int x = 0; x < MAP_COLS; x++) {
            if (map[y][x] == '@') map[y][x] = '.';
        }
    }
    if (initial_pos_x >= 0 && initial_pos_x < MAP_COLS &&
        initial_pos_y >= 0 && initial_pos_y < MAP_ROWS) {
        map[initial_pos_y][initial_pos_x] = '@';
    }
    save_map(); 
}

void move(char d) {
    int pos_x = get_piece_state_int("fuzzpet", "pos_x");
    int pos_y = get_piece_state_int("fuzzpet", "pos_y");
    int nx = pos_x, ny = pos_y;
    
    if (d == 'w') ny--; else if (d == 's') ny++; else if (d == 'a') nx--; else if (d == 'd') nx++;
    
    if (nx < 0 || nx >= MAP_COLS || ny < 0 || ny >= MAP_ROWS) return;
    if (map[ny][nx] == '#' || map[ny][nx] == 'R') return;
    
    if (map[ny][nx] == 'T') {
        int hap = get_piece_state_int("fuzzpet", "happiness");
        set_piece_state_int("fuzzpet", "happiness", (hap < 90) ? hap + 10 : 100);
        log_master("EventFire", "treasure_found", "fuzzpet", "fuzzpet_app");
    }
    
    map[pos_y][pos_x] = '.';
    pos_x = nx; pos_y = ny;
    map[pos_y][pos_x] = '@';
    
    set_piece_state_int("fuzzpet", "pos_x", pos_x);
    set_piece_state_int("fuzzpet", "pos_y", pos_y);
    
    int hunger = get_piece_state_int("fuzzpet", "hunger");
    int energy = get_piece_state_int("fuzzpet", "energy");
    set_piece_state_int("fuzzpet", "hunger", (hunger < 100) ? hunger + 1 : 100);
    set_piece_state_int("fuzzpet", "energy", (energy > 0) ? energy - 1 : 0);
    
    system("./pieces/apps/fuzzpet_app/clock/plugins/+x/increment_time.+x > /dev/null 2>&1");
    system("./pieces/apps/fuzzpet_app/clock/plugins/+x/increment_turn.+x > /dev/null 2>&1");
    
    strcpy(current_response_key, "moved");
    save_map();
    log_master("EventFire", "moved", "fuzzpet", "fuzzpet_app");
}

void process_key(int key) {
    time_t rawtime;
    struct tm *timeinfo;
    char timestamp[100];
    time(&rawtime);
    timeinfo = localtime(&rawtime);
    strftime(timestamp, sizeof(timestamp), "%Y-%m-%d %H:%M:%S", timeinfo);
    
    if (key >= 32 && key <= 126) snprintf(last_key_display, sizeof(last_key_display), "%d ('%c')", key, (char)key);
    else if (key == 10 || key == 13) strcpy(last_key_display, "ENTER");
    else if (key == 27) strcpy(last_key_display, "ESC");
    else snprintf(last_key_display, sizeof(last_key_display), "CODE %d", key);

    if (key == 'w' || key == 'W' || key == K_UP) move('w');
    else if (key == 's' || key == 'S' || key == K_DOWN) move('s');
    else if (key == 'a' || key == 'A' || key == K_LEFT) move('a');
    else if (key == 'd' || key == 'D' || key == K_RIGHT) move('d');
    else if (key == '2') { // Feed Pet (Aligned to UI Index 2)
        int h = get_piece_state_int("fuzzpet", "hunger");
        set_piece_state_int("fuzzpet", "hunger", (h > 20) ? h - 20 : 0);
        log_master("ResponseRequest", "fed", "fuzzpet", "fuzzpet_app");
        strcpy(current_response_key, "fed");
    }
    else if (key == '3') { // Play (Aligned to UI Index 3)
        int h = get_piece_state_int("fuzzpet", "happiness");
        set_piece_state_int("fuzzpet", "happiness", (h < 90) ? h + 10 : 100);
        log_master("ResponseRequest", "played", "fuzzpet", "fuzzpet_app");
        strcpy(current_response_key, "played");
    }
    else if (key == '4') { // Sleep (Aligned to UI Index 4)
        int e = get_piece_state_int("fuzzpet", "energy");
        set_piece_state_int("fuzzpet", "energy", (e < 70) ? e + 30 : 100);
        log_master("ResponseRequest", "slept", "fuzzpet", "fuzzpet_app");
        strcpy(current_response_key, "slept");
    }
    else if (key == '5') { 
        int l = get_piece_state_int("fuzzpet", "level");
        set_piece_state_int("fuzzpet", "level", l + 1);
        log_master("ResponseRequest", "evolved", "fuzzpet", "fuzzpet_app");
        strcpy(current_response_key, "evolved");
    }
    else if (key == '6') { 
        system("./pieces/apps/fuzzpet_app/clock/plugins/+x/end_turn.+x > /dev/null 2>&1");
        log_master("EventFire", "turn_end", "clock", "fuzzpet_app");
        strcpy(current_response_key, "default");
    }
    else {
        strcpy(current_response_key, "default");
    }

    write_sovereign_view();
}

void* input_thread(void* arg) {
    long last_pos = 0;
    FILE *h_init = fopen("pieces/apps/fuzzpet_app/fuzzpet/history.txt", "r");
    if (h_init) {
        fseek(h_init, 0, SEEK_END);
        last_pos = ftell(h_init);
        fclose(h_init);
    }
    
    while (1) {
        FILE *hf = fopen("pieces/apps/fuzzpet_app/fuzzpet/history.txt", "r");
        if (hf) {
            struct stat st;
            if (fstat(fileno(hf), &st) == 0) {
                if (st.st_size > last_pos) {
                    fseek(hf, last_pos, SEEK_SET);
                    int key;
                    while (fscanf(hf, "%d", &key) == 1) {
                        process_key(key);
                    }
                    last_pos = ftell(hf);
                }
            }
            fclose(hf);
        }
        usleep(16667); 
    }
    return NULL;
}

int main() {
    load_map();
    write_sovereign_view(); // HEARTBEAT: Initial frame write
    pthread_t t;
    pthread_create(&t, NULL, input_thread, NULL);
    while (1) {
        usleep(16667); 
    }
    return 0;
}

File: @.PTM.℗™/pmo-gh/1.PIECE.MOD.OS_c&ph_04.01/!.PMO_py_00.01/pieces/apps/fuzzpet_app/fuzzpet/plugins/fuzzpet.py
#!/usr/bin/env python3
"""
FuzzPet Module - Sovereign Stage (Python Implementation)
Standardized for PMO v1.4.0 (Option Z & Single Pulse).
Produces view.txt and syncs state.txt mirror for the OS Theater.
"""

import os
import sys
import time
import subprocess
import threading
from datetime import datetime

# Import TPM Utils
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../../../locations'))
import path_utils
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../../../master_ledger/plugins'))
from piece_manager import piece_manager

# Constants
MAP_ROWS = 10
MAP_COLS = 20

class FuzzPetStage:
    def __init__(self):
        self.state = {
            'name': 'Fuzzball',
            'hunger': 50,
            'happiness': 75,
            'energy': 100,
            'level': 1,
            'pos_x': 5,
            'pos_y': 5,
            'status': 'active'
        }
        self.map = []
        self.current_response_key = "default"
        self.last_key_display = "None"
        self.load_initial_data()

    def load_initial_data(self):
        """MIRROR C VERSION: Load state and map, place pet at correct position."""
        # Load state from DNA (.pdl mirror)
        dna = piece_manager.read_piece_state('fuzzpet')
        if dna: self.state.update(dna)
        
        # Load map - MIRROR C: load_map() clears old '@' and places at initial position
        self.map = piece_manager.load_map()
        
        # Ensure pet '@' is at the correct position from state
        # C version does this in load_map() by reading pos_x/pos_y from state
        pos_x = self.state.get('pos_x', 5)
        pos_y = self.state.get('pos_y', 5)
        
        # Clear any existing '@' from map
        for y in range(len(self.map)):
            if isinstance(self.map[y], list):
                for x in range(len(self.map[y])):
                    if self.map[y][x] == '@':
                        self.map[y][x] = '.'
            else:
                row = list(self.map[y])
                if '@' in row:
                    row[row.index('@')] = '.'
                    self.map[y] = "".join(row)
        
        # Place '@' at current position
        if 0 <= pos_y < len(self.map):
            if isinstance(self.map[pos_y], list):
                if 0 <= pos_x < len(self.map[pos_y]):
                    self.map[pos_y][pos_x] = '@'
            else:
                row = list(self.map[pos_y])
                if 0 <= pos_x < len(row):
                    row[pos_x] = '@'
                self.map[pos_y] = "".join(row)

    def sync_mirror(self):
        """DNA SYNC: Update piece.txt (Audit) and state.txt (High-speed Render)."""
        # 1. Update DNA (.pdl mirror via piece_manager)
        piece_manager.write_piece_state('fuzzpet', self.state)
        
        # 2. Update High-Speed Mirror (state.txt) for Parser
        mirror_path = path_utils.get_piece_path('fuzzpet', 'state.txt')
        if mirror_path:
            with open(mirror_path, 'w') as f:
                for k, v in self.state.items():
                    f.write(f"{k}={v}\n")
                f.write(f"response_key={self.current_response_key}\n")

    def write_sovereign_view(self):
        """MIRROR C VERSION EXACTLY."""
        # 1. Sync the mirror first so others see the data (Parser relies on this!)
        self.sync_mirror()

        # 2. Get Last Response (Audit Reflection) - C uses popen to piece_manager
        response = "..."
        responses = piece_manager.get_piece_responses('fuzzpet')
        if self.current_response_key in responses:
            response = responses[self.current_response_key]

        # 3. Compose Sovereign View (The "Stage") - EXACTLY LIKE C
        view_path = path_utils.get_piece_path('fuzzpet', 'view.txt')
        if not view_path: return
        
        with open(view_path, 'w') as fp:
            for y in range(MAP_ROWS):
                # C: fprintf(fp, "║  %-20s                                     ║\n", map[y]);
                # Handle both list of chars and string
                if y < len(self.map):
                    if isinstance(self.map[y], list):
                        map_line = "".join(self.map[y])
                    else:
                        map_line = self.map[y]
                else:
                    map_line = ""
                fp.write(f"║  {map_line:<57} ║\n")
            
            fp.write("╠═══════════════════════════════════════════════════════════╣\n")
            fp.write(f"║  [FUZZPET]: {response[:45]:<45} ║\n")
            fp.write(f"║  DEBUG [Last Key]: {self.last_key_display:<38} ║\n")
        
        # 4. SIGNALING NODE: Append to local marker to wake up the Parser (Option Z)
        pulse_path = path_utils.get_piece_path('fuzzpet', 'view_changed.txt')
        if pulse_path:
            with open(pulse_path, 'a') as f:
                f.write('X\n')

    def move(self, direction):
        """MIRROR C VERSION: Update map array directly."""
        dx, dy = 0, 0
        if direction == 'w': dy = -1
        elif direction == 's': dy = 1
        elif direction == 'a': dx = -1
        elif direction == 'd': dx = 1
        
        pos_x = self.state['pos_x']
        pos_y = self.state['pos_y']
        nx, ny = pos_x + dx, pos_y + dy
        
        if 0 <= nx < MAP_COLS and 0 <= ny < MAP_ROWS:
            # Check for walls (mirror C)
            if ny < len(self.map) and nx < len(self.map[ny]):
                if self.map[ny][nx] in ['#', 'R']:
                    return
            
            # Check for treasure (mirror C)
            if ny < len(self.map) and nx < len(self.map[ny]):
                if self.map[ny][nx] == 'T':
                    self.state['happiness'] = min(100, self.state['happiness'] + 10)
                    piece_manager.append_to_master_ledger("EventFire", "treasure_found", "fuzzpet")
            
            # Update map - MIRROR C: map[pos_y][pos_x] = '.'; map[pos_y][pos_x] = '@';
            if pos_y < len(self.map):
                row = list(self.map[pos_y])
                if 0 <= pos_x < len(row):
                    row[pos_x] = '.'
                self.map[pos_y] = "".join(row)
            
            self.state['pos_x'] = nx
            self.state['pos_y'] = ny
            
            if ny < len(self.map):
                row = list(self.map[ny])
                if 0 <= nx < len(row):
                    row[nx] = '@'
                self.map[ny] = "".join(row)
            
            # Update stats (mirror C)
            self.state['hunger'] = min(100, self.state['hunger'] + 1)
            self.state['energy'] = max(0, self.state['energy'] - 1)
            self.current_response_key = "moved"
            
            # Save map (mirror C)
            piece_manager.save_map(self.map)
            
            # Tick Clock (via plugins)
            clock_inc_time = path_utils.get_piece_path('clock', 'plugins/increment_time.py')
            clock_inc_turn = path_utils.get_piece_path('clock', 'plugins/increment_turn.py')
            if clock_inc_time:
                subprocess.run([sys.executable, clock_inc_time], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
            if clock_inc_turn:
                subprocess.run([sys.executable, clock_inc_turn], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
            
            piece_manager.append_to_master_ledger("EventFire", "moved", "fuzzpet")

    def process_key(self, key):
        # Debug display
        if 32 <= key <= 126: self.last_key_display = f"{key} ('{chr(key)}')"
        else: self.last_key_display = f"CODE {key}"

        char = chr(key) if 32 <= key <= 126 else ""
        
        if char in ['w', 'W']: self.move('w')
        elif char in ['s', 'S']: self.move('s')
        elif char in ['a', 'A']: self.move('a')
        elif char in ['d', 'D']: self.move('d')
        elif char == '2': # FEED (UI Index 2)
            self.state['hunger'] = max(0, self.state['hunger'] - 20)
            self.current_response_key = "fed"
            piece_manager.append_to_master_ledger("ResponseRequest", "fed", "fuzzpet")
        elif char == '3': # PLAY (UI Index 3)
            self.state['happiness'] = min(100, self.state['happiness'] + 10)
            self.current_response_key = "played"
            piece_manager.append_to_master_ledger("ResponseRequest", "played", "fuzzpet")
        elif char == '4': # SLEEP (UI Index 4)
            self.state['energy'] = min(100, self.state['energy'] + 30)
            self.current_response_key = "slept"
            piece_manager.append_to_master_ledger("ResponseRequest", "slept", "fuzzpet")
        elif char == '5': # EVOLVE
            self.state['level'] += 1
            self.current_response_key = "evolved"
            piece_manager.append_to_master_ledger("ResponseRequest", "evolved", "fuzzpet")
        elif char == '6': # END TURN
            clock_end = path_utils.get_piece_path('clock', 'plugins/end_turn.py')
            if clock_end:
                subprocess.run([sys.executable, clock_end], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
            self.current_response_key = "default"
            piece_manager.append_to_master_ledger("EventFire", "turn_end", "clock")
        else:
            self.current_response_key = "default"

        self.write_sovereign_view()

def watch_input(stage):
    history_path = path_utils.get_piece_path('fuzzpet', 'history.txt')
    if not history_path: return
    
    # Ensure exists
    os.makedirs(os.path.dirname(history_path), exist_ok=True)
    if not os.path.exists(history_path): open(history_path, 'w').close()
    
    last_pos = os.path.getsize(history_path)
    while True:
        try:
            curr_size = os.path.getsize(history_path)
            if curr_size < last_pos: last_pos = 0 # Truncated
            
            if curr_size > last_pos:
                with open(history_path, 'r') as f:
                    f.seek(last_pos)
                    for line in f:
                        if line.strip().isdigit():
                            stage.process_key(int(line.strip()))
                    last_pos = f.tell()
            time.sleep(0.016) # 60 FPS Polling
        except:
            time.sleep(0.1)

if __name__ == "__main__":
    stage = FuzzPetStage()
    # Initial Frame - C calls write_sovereign_view() at startup
    stage.write_sovereign_view()
    print("[fuzzpet] Sovereign Stage active. Monitoring history.txt...", file=sys.stderr)
    watch_input(stage)
